org.scala-ide.sdt.debug.tests/src/org/scalaide/debug/internal/launching/LaunchUtils.scala

```diff
@@ -0,0 +1,56 @@
+package org.scalaide.debug.internal.launching
+
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+
+import org.eclipse.core.resources.IProject
+import org.eclipse.core.resources.IncrementalProjectBuilder
+import org.eclipse.core.runtime.NullProgressMonitor
+import org.eclipse.debug.core.DebugPlugin
+import org.eclipse.debug.core.ILaunch
+import org.eclipse.debug.core.ILaunchConfiguration
+import org.eclipse.debug.core.ILaunchesListener2
+
+/**
+ * Used for launching application.
+ */
+trait LaunchUtils {
+  /** Points to launch configuration file. */
+  val launchConfigurationName: String
+
+  private val DefaultMonitor = new NullProgressMonitor
+
+  /** Create a launch listener for launchTerminated events on a launch of the given launchConfiguration. */
+  def onLaunchTerminates(f: () => Unit) = new ILaunchesListener2() {
+    override def launchesTerminated(launches: Array[ILaunch]): Unit = {
+      if (launches.exists(_.getLaunchConfiguration.getName == launchConfigurationName)) {
+        f()
+      }
+    }
+    override def launchesAdded(launches: Array[ILaunch]): Unit = {}
+    override def launchesRemoved(launches: Array[ILaunch]): Unit = {}
+    override def launchesChanged(launches: Array[ILaunch]): Unit = {}
+  }
+
+  /** Cleans and incrementally builds projects */
+  def cleanBuild(projects: IProject*): Unit = projects.foreach { project =>
+    project.build(IncrementalProjectBuilder.CLEAN_BUILD, DefaultMonitor)
+    project.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, DefaultMonitor)
+  }
+
+  private def launchConfiguration(project: IProject): ILaunchConfiguration =
+    DebugPlugin.getDefault.getLaunchManager.getLaunchConfiguration(project.getFile(launchConfigurationName + ".launch"))
+
+  def whenApplicationWasLaunchedFor(project: IProject, inMode: String)(inThatCase: => Unit): Unit = {
+    val latch = new CountDownLatch(1)
+    DebugPlugin.getDefault.getLaunchManager.addLaunchListener(onLaunchTerminates(latch.countDown))
+    val lc = launchConfiguration(project)
+    val launch = lc.launch(inMode, DefaultMonitor)
+    val timeout = if (launch.canTerminate) 10 else 60
```

Comment:
digging in code you find that both timeouts are taken from existing tests
implementation (so I cannot clearly explain why is 10 and 60 secs), I've
just moved them here. Not every launch can terminate, in such case callback
'onTerminate' is not called.

2015-05-27 14:48 GMT+02:00 Simon Schäfer <notifications@github.com>:

> In
> org.scala-ide.sdt.debug.tests/src/org/scalaide/debug/internal/launching/LaunchUtils.scala
> <https://github.com/scala-ide/scala-ide/pull/955#discussion_r31129180>:
>
> > +
> > +  /** Cleans and incrementally builds projects */
> > +  def cleanBuild(projects: IProject*): Unit = projects.foreach { project =>
> > +    project.build(IncrementalProjectBuilder.CLEAN_BUILD, DefaultMonitor)
> > +    project.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, DefaultMonitor)
> > +  }
> > +
> > +  private def launchConfiguration(project: IProject): ILaunchConfiguration =
> > +    DebugPlugin.getDefault.getLaunchManager.getLaunchConfiguration(project.getFile(launchConfigurationName + ".launch"))
> > +
> > +  def whenApplicationWasLaunchedFor(project: IProject, inMode: String)(inThatCase: => Unit): Unit = {
> > +    val latch = new CountDownLatch(1)
> > +    DebugPlugin.getDefault.getLaunchManager.addLaunchListener(onLaunchTerminates(latch.countDown))
> > +    val lc = launchConfiguration(project)
> > +    val launch = lc.launch(inMode, DefaultMonitor)
> > +    val timeout = if (launch.canTerminate) 10 else 60
>
> What do this timeouts mean? How did you come up with them?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scala-ide/scala-ide/pull/955/files#r31129180>.
>